worker/download.go

```diff
@@ -0,0 +1,1088 @@
+	var sID slabID
+	frand.Read(sID[:])
+
+	// add slab to ongoing downloads
+	mgr.mu.Lock()
+	mgr.ongoing[sID] = struct{}{}
+	mgr.mu.Unlock()
+
+	// prepare a function to remove it from the ongoing downloads
+	finishFn := func() {
+		mgr.mu.Lock()
+		delete(mgr.ongoing, sID)
+		mgr.mu.Unlock()
+	}
+
+	// calculate the offset and length
+	offset, length := slice.SectorRegion()
+
+	// build sector info
+	hostToSectors := make(map[types.PublicKey][]sectorInfo)
+	for sI, s := range slice.Shards {
+		hostToSectors[s.Host] = append(hostToSectors[s.Host], sectorInfo{s, sI})
+	}
+
+	// create slab download
+	return &slabDownload{
+		mgr: mgr,
+
+		dID:       dID,
+		sID:       sID,
+		created:   time.Now(),
+		index:     slabIndex,
+		minShards: int(slice.MinShards),
+		offset:    offset,
+		length:    length,
+
+		hostToSectors: hostToSectors,
+		used:          make(map[types.PublicKey]struct{}),
+
+		sectors: make([][]byte, len(slice.Shards)),
+	}, finishFn
+}
+
+func (mgr *downloadManager) ongoingDownloads() int {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	return len(mgr.ongoing)
+}
+
+func (mgr *downloadManager) downloadSlab(ctx context.Context, dID id, slice object.SlabSlice, index int, responseChan chan *slabDownloadResponse, nextSlabChan chan struct{}) {
+	// add tracing
+	ctx, span := tracing.Tracer.Start(ctx, "downloadSlab")
+	defer span.End()
+
+	// prepare the download
+	slab, finishFn := mgr.newSlabDownload(ctx, dID, slice, index)
+	defer finishFn()
+
+	// download shards
+	resp := &slabDownloadResponse{index: index}
+	resp.shards, resp.err = slab.downloadShards(ctx, nextSlabChan)
+
+	// send the response
+	select {
+	case <-ctx.Done():
+	case responseChan <- resp:
+	}
+}
+
+func (d *downloader) stats() downloaderStats {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	return downloaderStats{
+		avgSpeedMBPS: d.statsDownloadSpeedBytesPerMS.Average() * 0.008,
+		healthy:      d.consecutiveFailures == 0,
+		numDownloads: d.numDownloads,
+	}
+}
+
+func (d *downloader) isStopped() bool {
+	select {
+	case <-d.stopChan:
+		return true
+	default:
+	}
+	return false
+}
+
+func (d *downloader) processQueue(hp hostProvider) {
+outer:
+	for {
+		// wait for work
+		select {
+		case <-d.signalWorkChan:
+		case <-d.stopChan:
+			return
+		}
+
+		for {
+			// check if we are stopped
+			if d.isStopped() {
+				return
+			}
+
+			// fill batch
+			var batch []*sectorDownloadReq
+			for len(batch) < maxConcurrentSectorsPerHost {
+				if req := d.pop(); req == nil {
+					break
+				} else if req.done() {
+					continue
+				} else {
+					batch = append(batch, req)
+				}
+			}
+			if len(batch) == 0 {
+				continue outer
+			}
+
+			fmt.Printf("DEBUG PJ: downloader %v is processing batch of size %v\n", d.host.HostKey(), len(batch))
+
+			var mu sync.Mutex
+			var completed int
+			var failed int
+			var first time.Time
+
+			// launch workers
+			var wg sync.WaitGroup
+			jobsChan := make(chan *sectorDownloadReq)
+			for i := 0; i < len(batch); i++ {
```

Comment:
Maybe this method can be cleaned up a bit to reduce nesting. E.g. have a method `processBatch` that contains some logic and then `processJob` or something.